sklearn/utils/extmath.py

```diff
@@ -1046,12 +1053,29 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
     -------
     result
         The output of the accumulator function passed to this function.
+
+    Notes
+    -----
+    When using array-api support, the accumulator function will upcast floating-point
+    arguments to the maximum precision possible for the array namespace and device.
+    This is usually float64, but may be float32 for some namespace/device pairs.
     """
-    if np.issubdtype(x.dtype, np.floating) and x.dtype.itemsize < 8:
-        result = op(x, *args, **kwargs, dtype=np.float64)
-    else:
-        result = op(x, *args, **kwargs)
-    return result
+    xp, _, x_device = get_namespace_and_device(x)
+    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
+        # We need to upcast. Some ops support this natively; others don't.
+        target_dtype = max_precision_float_dtype(xp, device=x_device)
+
+        def convert_(arr):
+            return xp.astype(arr, target_dtype, copy=False)
+
+        try:
+            return op(x, *args, **kwargs, dtype=target_dtype)
+        except TypeError:
```

Comment:
Instead of using a `try/except TypeError`, could you please inspect the signature of `op` to check for the presence of a parameter named `"dtype"`? That would make the code more explicit I think.